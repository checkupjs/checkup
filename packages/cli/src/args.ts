export const pathArg = {
  name: 'paths',
  description:
    'The paths or globs that checkup will operate on. If no paths or globs are provided, checkup will run on the entire directory beginning at --cwd.',
};
