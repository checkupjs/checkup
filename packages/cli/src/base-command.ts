import { CheckupError } from '@checkup/core';
import { Command } from '@oclif/command';
import { red } from 'chalk';

export abstract class BaseCommand extends Command {
  extendedError(
    error: string | Error | CheckupError,
    options: { code?: string; exit?: number | false } = {}
  ) {
    if (typeof error === 'string') {
      error = new Error(error);
    }

    if (options.exit === false) {
      console.error(
        error instanceof CheckupError ? error.render() : `\n${red('Error')} ${error.message}`
      );
    } else {
      throw error;
    }
  }
}
