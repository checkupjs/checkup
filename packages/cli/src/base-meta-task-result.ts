import { TaskIdentifier } from '@checkup/core';

export default abstract class BaseMetaTaskResult {
  meta: TaskIdentifier;

  constructor(meta: TaskIdentifier) {
    this.meta = meta;
  }
}
