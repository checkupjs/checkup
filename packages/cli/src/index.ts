import * as path from 'path';

import { Command, flags } from '@oclif/command';
import {
  TaskConstructor,
  TaskList,
  TaskResult,
  getPackageJson,
  getRegisteredTasks,
  ui,
} from '@checkup/core';

function mergeTaskResults(taskResults: TaskResult[]) {
  let mergedResults = {};

  taskResults.forEach(taskResult => {
    mergedResults = Object.assign(mergedResults, taskResult.toJson());
  });

  return mergedResults;
}

class Checkup extends Command {
  static description = 'A CLI that provides health check information about your project';

  static args = [
    {
      name: 'path',
      required: true,
      description: 'The path referring to the root directory that Checkup will run in',
      default: '.',
    },
  ];

  static flags = {
    version: flags.version({ char: 'v' }),
    help: flags.help({ char: 'h' }),
    force: flags.boolean({ char: 'f' }),
    silent: flags.boolean({ char: 's' }),
    json: flags.boolean(),
    task: flags.string({ char: 't' }),
  };

  async run() {
    let { args, flags } = this.parse(Checkup);
    let registeredTasks: Map<string, TaskConstructor>;

    try {
      getPackageJson(args.path);
    } catch (e) {
      this.error(
        `The ${path.resolve(
          args.path
        )} directory found through the 'path' option does not contain a package.json file. You must run checkup in a directory with a package.json file.`,
        e
      );
    }

    await this.config.runHook('register-tasks', {});

    registeredTasks = getRegisteredTasks();

    let taskResults;
    let tasksToBeRun = new TaskList();

    if (flags.task !== undefined) {
      let task = registeredTasks.get(flags.task);

      if (task !== undefined) {
        tasksToBeRun.addTask(task, args);
      }
    } else {
      tasksToBeRun.addTasks(Array.from(registeredTasks.values()), args);
    }

    ui.action.start('Checking up on your project');

    taskResults = await tasksToBeRun.runTasks();

    if (!flags.silent) {
      if (flags.json) {
        ui.styledJSON(mergeTaskResults(taskResults));
      } else {
        taskResults.forEach(taskResult => taskResult.toConsole());
      }
    }

    ui.action.stop();
  }
}

export = Checkup;
