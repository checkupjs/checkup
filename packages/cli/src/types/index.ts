export default {};

export const enum ProjectType {
  App = 'application',
  Addon = 'addon',
  Engine = 'engine',
  Unknown = 'unknown',
}

export const enum TestType {
  Application = 'application',
  Container = 'container',
  Rendering = 'rendering',
  Unit = 'unit',
}

export interface Task {
  run: () => Promise<TaskResult>;
}

export interface TaskConstructor {
  new (): Task;
}

export interface TaskItemData {
  type: string;
  data: string[];
  total: number;
}

export interface TaskResult {
  toConsole: () => void;
  toJson: () => {};
}

export type SearchPatterns = Record<string, string[]>;
