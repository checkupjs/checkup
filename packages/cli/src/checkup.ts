import cli from 'cli-ux';
import { ITaskConstructor, ITaskResult } from './types';
import { getTaskByName } from './utils/default-tasks';
import TaskList from './task-list';
import * as DefaultTasks from './tasks';
import ResultWriter from './utils/result-writer';
import Clock from './utils/clock';

const DEFAULT_TASKS = <ITaskConstructor[]>(
  Object.values(DefaultTasks).filter(x => typeof x == 'function')
);

/**
 * @class Checkup
 *
 * The entry point for invoking all checkup tasks.
 */
export default class Checkup {
  args: any;
  flags: any;
  defaultTasks: ITaskConstructor[];

  /**
   *
   * @param project {IProject} the project model that is instantiated as part of ember-cli.
   * @param ui {IUserInterface} the UI model that is instantiated as part of ember-cli.
   */
  constructor(args: any, flags: any, tasks: ITaskConstructor[] = DEFAULT_TASKS) {
    this.args = args;
    this.flags = flags;
    this.defaultTasks = tasks;
  }

  /**
   * @method run
   *
   * Gathers and runs all tasks associated with checking up on an Ember repo.
   */
  async run(): Promise<ITaskResult[]> {
    let clock = new Clock();
    let tasks = new TaskList();

    if (this.flags.task !== undefined) {
      let task = getTaskByName(this.flags.task);

      tasks.addTask(task);
    } else {
      tasks.addTasks(this.defaultTasks);
    }

    clock.start();

    cli.action.start('Hang tight while we check up on your Ember project');

    let taskResults = await tasks.runTasks();

    clock.stop();

    cli.action.stop();

    if (!this.flags.silent) {
      let writer = new ResultWriter(taskResults);

      if (this.flags.json) {
        console.log(JSON.stringify(writer.toJson(), null, 2));
      } else {
        writer.toConsole();
      }
      writer.writeDuration(clock.duration);
    }

    return taskResults;
  }
}
