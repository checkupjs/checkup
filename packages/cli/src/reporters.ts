import {
  ReporterType,
  TaskResult,
  ui,
  Category,
  Priority,
  ReportComponentType,
  UIResultData,
  UIReportData,
} from '@checkup/core';

import { MetaTaskResult } from './types';
import { RunFlags } from './commands/run';
import { generateHTMLReport } from './helpers/ui-report';

export function _transformHTMLResults(
  metaTaskResults: MetaTaskResult[],
  pluginTaskResults: TaskResult[]
): UIReportData {
  let mergedResults: UIResultData = {
    [Category.Insights]: {
      [Priority.High]: [],
      [Priority.Medium]: [],
      [Priority.Low]: [],
    },
    [Category.Migrations]: {
      [Priority.High]: [],
      [Priority.Medium]: [],
      [Priority.Low]: [],
    },
    [Category.Recommendations]: {
      [Priority.High]: [],
      [Priority.Medium]: [],
      [Priority.Low]: [],
    },
  };
  let requiresChart = false;

  pluginTaskResults
    .flatMap((result) => result.html())
    .forEach((taskResult) => {
      if (taskResult) {
        let { category, priority } = taskResult.meta.taskClassification;

        mergedResults[category][priority].push(taskResult);

        if (taskResult.componentType === ReportComponentType.PieChart) {
          requiresChart = true;
        }
      }
    });

  return {
    meta: Object.assign({}, ...metaTaskResults.map((result) => result.json())),
    results: mergedResults,
    requiresChart,
  };
}

export function _transformJsonResults(
  metaTaskResults: MetaTaskResult[],
  pluginTaskResults: TaskResult[]
) {
  let transformedResult = {
    meta: Object.assign({}, ...metaTaskResults.map((result) => result.json())),
    results: pluginTaskResults.map((result) => result.json()),
  };

  return transformedResult;
}

export function getReporter(
  flags: RunFlags,
  metaTaskResults: MetaTaskResult[],
  pluginTaskResults: TaskResult[]
) {
  switch (flags.reporter) {
    case ReporterType.stdout:
      return async () => {
        if (!flags.silent) {
          metaTaskResults.forEach((taskResult) => taskResult.stdout());
          pluginTaskResults.forEach((taskResult) => taskResult.stdout());
        }
      };
    case ReporterType.json:
      return async () => {
        let resultJson = _transformJsonResults(metaTaskResults, pluginTaskResults);
        ui.styledJSON(resultJson);
      };
    case ReporterType.html:
      return async () => {
        let resultsForPdf = _transformHTMLResults(metaTaskResults, pluginTaskResults);
        let reportPath = await generateHTMLReport(flags.reportOutputPath, resultsForPdf);

        ui.log(reportPath);
      };
    default:
      return async () => {};
  }
}
