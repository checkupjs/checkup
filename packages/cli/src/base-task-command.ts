import { Result, Log, Invocation } from 'sarif';
import {
  RunFlags,
  TaskError,
  Action,
  CheckupConfig,
  Task,
  TaskContext,
  ui,
  CheckupError,
  getRegisteredActions,
  getConfigPathFromOptions,
  getConfigPath,
  readConfig,
  registerParser,
  registerActions,
  registerTaskReporter,
  loadPlugins,
  getRegisteredParsers,
  FilePathArray,
  getFilePathsAsync,
  buildNotificationsFromTaskErrors,
} from '@checkup/core';
import { BaseCommand } from './base-command';
import TaskList from './task-list';
import { getLog } from './get-log';
import { TASK_ERRORS } from './task-errors';
import { getPackageJson } from './utils/get-package-json';
import { getReporter } from './reporters/get-reporter';
import { reportAvailableTasks } from './reporters/available-task-reporter';

let __tasksForTesting: Set<Task> = new Set<Task>();

export function _registerTaskForTesting(task: Task) {
  __tasksForTesting.add(task);
}
export function _resetTasksForTesting() {
  __tasksForTesting = new Set<Task>();
}

export abstract class BaseTaskCommand extends BaseCommand {
  runArgs!: string[];
  runFlags!: RunFlags;
  pluginTasks: TaskList = new TaskList();
  pluginTaskResults: Result[] = [];
  pluginTaskErrors: TaskError[] = [];
  startTime: string = '';
  actions: Action[] = [];
  checkupConfig!: CheckupConfig;
  executedTasks!: Task[];
  cliModeEnabled: boolean = true;
  taskContext!: TaskContext;

  get taskFilterType() {
    if (this.runFlags.task !== undefined) {
      return 'task';
    } else if (this.runFlags.category !== undefined) {
      return 'category';
    } else if (this.runFlags.group !== undefined) {
      return 'group';
    }

    return '';
  }

  public async run() {
    await this.loadConfig();

    await this.register();

    if (this.runFlags['list-tasks']) {
      this.printAvailableTasks();
    } else {
      if (this.cliModeEnabled) {
        ui.action.start('Checking up on your project');
      }

      await this.runTasks();
      await this.runActions();

      let log: Log = await getLog(
        this.taskContext,
        this.pluginTaskResults,
        this.actions,
        this.getInvocation(this.pluginTaskErrors),
        this.pluginTasks,
        this.executedTasks
      );

      if (this.cliModeEnabled) {
        this.report(log);
        ui.action.stop();
      }

      return log;
    }
  }

  private async runTasks() {
    if (this.taskFilterType) {
      let { tasksFound, tasksNotFound } = this.findTasks();

      if (tasksFound.length > 0) {
        [this.pluginTaskResults, this.pluginTaskErrors] = await this.pluginTasks.runTasks(
          tasksFound
        );
      }

      if (tasksNotFound.length > 0) {
        let error = TASK_ERRORS.get(this.taskFilterType)!;

        this.extendedError(new CheckupError(error.message(tasksNotFound), error.callToAction));
        ui.action.stop();
      }

      this.executedTasks = tasksFound;
    } else {
      [this.pluginTaskResults, this.pluginTaskErrors] = await this.pluginTasks.runTasks();
      this.executedTasks = this.pluginTasks.getTasks();
    }
  }

  private findTasks() {
    if (this.runFlags.task !== undefined) {
      return this.pluginTasks.findAllByTaskName(...this.runFlags.task);
    } else if (this.runFlags.category !== undefined) {
      return this.pluginTasks.findAllByCategory(...this.runFlags.category);
    } else if (this.runFlags.group !== undefined) {
      return this.pluginTasks.findAllByGroup(...this.runFlags.group);
    }

    return { tasksFound: [], tasksNotFound: [] };
  }

  private runActions() {
    return new Promise<void>((resolve, reject) => {
      let evaluators = getRegisteredActions();

      for (let [taskName, evaluator] of evaluators) {
        let task = this.pluginTasks.find(taskName);

        let taskResult = this.pluginTaskResults.filter((result: Result) => {
          return result.properties?.taskName === taskName;
        });

        if (task && taskResult.length > 0) {
          try {
            this.actions.push(...evaluator(taskResult, task.config));
          } catch (error) {
            reject(error);
          }
        }
      }

      resolve();
    });
  }

  private async loadConfig() {
    let configPath;

    try {
      configPath =
        (await getConfigPathFromOptions(this.runFlags.config)) || getConfigPath(this.runFlags.cwd);
      this.checkupConfig = readConfig(configPath);

      let plugins = await loadPlugins(this.checkupConfig.plugins, this.runFlags.cwd);

      this.config.plugins.push(...plugins);
    } catch (error) {
      this.extendedError(error);
    }
  }

  protected async register() {
    await this.config.runHook('register-parsers', {
      registerParser,
    });

    await this.config.runHook('register-actions', {
      registerActions,
    });

    await this.config.runHook('register-task-reporter', {
      registerTaskReporter,
    });

    // if excludePaths are provided both via the command line and config, the command line is prioritized
    let excludePaths = this.runFlags['exclude-paths'] || this.checkupConfig.excludePaths;

    this.taskContext = Object.freeze({
      cliArguments: this.runArgs,
      cliFlags: this.runFlags,
      parsers: getRegisteredParsers(),
      config: this.checkupConfig,
      pkg: getPackageJson(this.runFlags.cwd),
      paths: FilePathArray.from(
        await getFilePathsAsync(this.runFlags.cwd, this.runArgs, excludePaths)
      ) as FilePathArray,
    });

    this.registerTasks();

    __tasksForTesting.forEach((task: Task) => {
      this.pluginTasks.registerTask(task);
    });
  }

  protected abstract registerTasks(): void;

  private report(log: Log) {
    let generateReport = getReporter(this.runFlags);

    generateReport(log, this.runFlags);
  }

  private printAvailableTasks() {
    reportAvailableTasks(this.pluginTasks);
  }

  private getInvocation(errors: TaskError[]): Invocation {
    return {
      arguments: this.runArgs,
      executionSuccessful: true,
      endTimeUtc: new Date().toJSON(),
      environmentVariables: {
        cwd: this.runFlags.cwd,
        outputFile: this.runFlags['output-file'],
        format: this.runFlags.format,
      },
      toolExecutionNotifications: buildNotificationsFromTaskErrors(errors),
      startTimeUtc: this.startTime,
    };
  }
}
