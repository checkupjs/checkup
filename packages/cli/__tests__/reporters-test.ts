import { Category, Priority, ReporterType, TaskResult } from '@checkup/core';

import { MetaTaskResult } from '../src/types';
import MockMetaTaskResult from './__utils__/mock-meta-task-result';
import MockTaskResult from './__utils__/mock-task-result';
import { _transformResults } from '../src/reporters';

describe('_transformResults', () => {
  let metaTaskResults: MetaTaskResult[];
  let pluginTaskResults: TaskResult[];

  it('transforms meta and plugin results into correct format', () => {
    metaTaskResults = [
      new MockMetaTaskResult(
        {
          taskName: 'mock-meta-task-1',
          friendlyTaskName: 'Mock Meta Task 1',
        },
        {
          foo: true,
          bar: 'baz',
        }
      ),
      new MockMetaTaskResult(
        {
          taskName: 'mock-meta-task-2',
          friendlyTaskName: 'Mock Meta Task 2',
        },
        {
          blarg: false,
          bleek: { bork: 'bye!' },
        }
      ),
    ];

    pluginTaskResults = [
      new MockTaskResult(
        {
          taskName: 'mock-meta-task-6',
          friendlyTaskName: 'Mock Meta Task 6',
          taskClassification: {
            category: Category.Insights,
            priority: Priority.Low,
          },
        },
        {
          foo: true,
          bar: 'baz',
        }
      ),
      new MockTaskResult(
        {
          taskName: 'mock-meta-task-7',
          friendlyTaskName: 'Mock Meta Task 7',
          taskClassification: {
            category: Category.Migrations,
            priority: Priority.High,
          },
        },
        {
          foo: true,
          bar: 'baz',
        }
      ),
      new MockTaskResult(
        {
          taskName: 'mock-meta-task-3',
          friendlyTaskName: 'Mock Meta Task 3',
          taskClassification: {
            category: Category.Insights,
            priority: Priority.High,
          },
        },
        {
          foo: true,
          bar: 'baz',
        }
      ),
      new MockTaskResult(
        {
          taskName: 'mock-meta-task-5',
          friendlyTaskName: 'Mock Meta Task 5',
          taskClassification: {
            category: Category.Insights,
            priority: Priority.High,
          },
        },
        {
          foo: true,
          bar: 'baz',
        }
      ),
      new MockTaskResult(
        {
          taskName: 'mock-meta-task-8',
          friendlyTaskName: 'Mock Meta Task 8',
          taskClassification: {
            category: Category.Migrations,
            priority: Priority.Low,
          },
        },
        {
          foo: true,
          bar: 'baz',
        }
      ),
      new MockTaskResult(
        {
          taskName: 'mock-meta-task-4',
          friendlyTaskName: 'Mock Meta Task 4',
          taskClassification: {
            category: Category.Insights,
            priority: Priority.Medium,
          },
        },
        {
          foo: true,
          bar: 'baz',
        }
      ),
    ];

    let transformed = _transformResults(metaTaskResults, pluginTaskResults, ReporterType.json);

    expect(transformed).toMatchInlineSnapshot(`
      Object {
        "meta": Object {
          "mock-meta-task-1": Object {
            "bar": "baz",
            "foo": true,
          },
          "mock-meta-task-2": Object {
            "blarg": false,
            "bleek": Object {
              "bork": "bye!",
            },
          },
        },
        "results": Array [
          Object {
            "meta": Object {
              "friendlyTaskName": "Mock Meta Task 6",
              "taskClassification": Object {
                "category": "insights",
                "priority": "low",
              },
              "taskName": "mock-meta-task-6",
            },
            "result": Object {
              "bar": "baz",
              "foo": true,
            },
          },
          Object {
            "meta": Object {
              "friendlyTaskName": "Mock Meta Task 7",
              "taskClassification": Object {
                "category": "migrations",
                "priority": "high",
              },
              "taskName": "mock-meta-task-7",
            },
            "result": Object {
              "bar": "baz",
              "foo": true,
            },
          },
          Object {
            "meta": Object {
              "friendlyTaskName": "Mock Meta Task 3",
              "taskClassification": Object {
                "category": "insights",
                "priority": "high",
              },
              "taskName": "mock-meta-task-3",
            },
            "result": Object {
              "bar": "baz",
              "foo": true,
            },
          },
          Object {
            "meta": Object {
              "friendlyTaskName": "Mock Meta Task 5",
              "taskClassification": Object {
                "category": "insights",
                "priority": "high",
              },
              "taskName": "mock-meta-task-5",
            },
            "result": Object {
              "bar": "baz",
              "foo": true,
            },
          },
          Object {
            "meta": Object {
              "friendlyTaskName": "Mock Meta Task 8",
              "taskClassification": Object {
                "category": "migrations",
                "priority": "low",
              },
              "taskName": "mock-meta-task-8",
            },
            "result": Object {
              "bar": "baz",
              "foo": true,
            },
          },
          Object {
            "meta": Object {
              "friendlyTaskName": "Mock Meta Task 4",
              "taskClassification": Object {
                "category": "insights",
                "priority": "medium",
              },
              "taskName": "mock-meta-task-4",
            },
            "result": Object {
              "bar": "baz",
              "foo": true,
            },
          },
        ],
      }
    `);
  });
});
