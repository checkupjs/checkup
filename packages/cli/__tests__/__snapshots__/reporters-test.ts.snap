// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_transformHTMLResults transforms meta and plugin results into correct format when chart IS required 1`] = `
Object {
  "meta": Object {
    "mock-meta-task-1": Object {
      "bar": "baz",
      "foo": true,
    },
    "mock-meta-task-2": Object {
      "blarg": false,
      "bleek": Object {
        "bork": "bye!",
      },
    },
  },
  "requiresChart": true,
  "results": Object {
    "insights": Object {
      "high": Array [],
      "low": Array [],
      "medium": Array [
        PieChartData {
          "UUID": 100,
          "backgroundColors": Array [
            "\\"#4FD1C5\\"",
            "\\"#7F9CF5\\"",
            "\\"#FC8181\\"",
            "\\"#63B3ED\\"",
          ],
          "componentType": "pie-chart",
          "grade": "A",
          "labels": Array [
            "\\"blah\\"",
            "\\"blah\\"",
            "\\"black\\"",
            "\\"sheep\\"",
          ],
          "meta": Object {
            "friendlyTaskName": "Mock Task",
            "taskClassification": Object {
              "category": "insights",
              "priority": "medium",
            },
            "taskName": "mock-task",
          },
          "resultDescription": "this is a chart",
          "values": Array [
            33,
            23,
            13,
            3,
          ],
        },
      ],
    },
    "migrations": Object {
      "high": Array [],
      "low": Array [],
      "medium": Array [],
    },
    "recommendations": Object {
      "high": Array [],
      "low": Array [],
      "medium": Array [],
    },
  },
}
`;

exports[`_transformHTMLResults transforms meta and plugin results into correct format when chart is not required 1`] = `
Object {
  "meta": Object {
    "mock-meta-task-1": Object {
      "bar": "baz",
      "foo": true,
    },
    "mock-meta-task-2": Object {
      "blarg": false,
      "bleek": Object {
        "bork": "bye!",
      },
    },
  },
  "requiresChart": false,
  "results": Object {
    "insights": Object {
      "high": Array [
        NumericalCardData {
          "componentType": "numerical-card",
          "grade": "A",
          "meta": Object {
            "friendlyTaskName": "Mock Meta Task 3",
            "taskClassification": Object {
              "category": "insights",
              "priority": "high",
            },
            "taskName": "mock-meta-task-3",
          },
          "resultData": Object {
            "bar": "baz",
            "foo": true,
          },
          "resultDescription": "this is a description of your result",
          "resultHelp": undefined,
        },
        NumericalCardData {
          "componentType": "numerical-card",
          "grade": "A",
          "meta": Object {
            "friendlyTaskName": "Mock Meta Task 5",
            "taskClassification": Object {
              "category": "insights",
              "priority": "high",
            },
            "taskName": "mock-meta-task-5",
          },
          "resultData": Object {
            "bar": "baz",
            "foo": true,
          },
          "resultDescription": "this is a description of your result",
          "resultHelp": undefined,
        },
      ],
      "low": Array [
        NumericalCardData {
          "componentType": "numerical-card",
          "grade": "A",
          "meta": Object {
            "friendlyTaskName": "Mock Meta Task 6",
            "taskClassification": Object {
              "category": "insights",
              "priority": "low",
            },
            "taskName": "mock-meta-task-6",
          },
          "resultData": Object {
            "bar": "baz",
            "foo": true,
          },
          "resultDescription": "this is a description of your result",
          "resultHelp": undefined,
        },
      ],
      "medium": Array [
        NumericalCardData {
          "componentType": "numerical-card",
          "grade": "A",
          "meta": Object {
            "friendlyTaskName": "Mock Meta Task 4",
            "taskClassification": Object {
              "category": "insights",
              "priority": "medium",
            },
            "taskName": "mock-meta-task-4",
          },
          "resultData": Object {
            "bar": "baz",
            "foo": true,
          },
          "resultDescription": "this is a description of your result",
          "resultHelp": undefined,
        },
      ],
    },
    "migrations": Object {
      "high": Array [
        NumericalCardData {
          "componentType": "numerical-card",
          "grade": "A",
          "meta": Object {
            "friendlyTaskName": "Mock Meta Task 7",
            "taskClassification": Object {
              "category": "migrations",
              "priority": "high",
            },
            "taskName": "mock-meta-task-7",
          },
          "resultData": Object {
            "bar": "baz",
            "foo": true,
          },
          "resultDescription": "this is a description of your result",
          "resultHelp": undefined,
        },
      ],
      "low": Array [
        NumericalCardData {
          "componentType": "numerical-card",
          "grade": "A",
          "meta": Object {
            "friendlyTaskName": "Mock Meta Task 8",
            "taskClassification": Object {
              "category": "migrations",
              "priority": "low",
            },
            "taskName": "mock-meta-task-8",
          },
          "resultData": Object {
            "bar": "baz",
            "foo": true,
          },
          "resultDescription": "this is a description of your result",
          "resultHelp": undefined,
        },
      ],
      "medium": Array [],
    },
    "recommendations": Object {
      "high": Array [],
      "low": Array [],
      "medium": Array [],
    },
  },
}
`;
